va/vampic.go

```diff
@@ -0,0 +1,427 @@
+	}
+
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	responses := make(chan *response, remoteVACount)
+	for _, i := range rand.Perm(remoteVACount) {
+		go func(rva RemoteVA) {
+			res, err := op(subCtx, rva, req)
+			if err != nil {
+				responses <- &response{rva.Address, rva.Perspective, rva.RIR, res, err}
+				return
+			}
+			if res.GetPerspective() != rva.Perspective || res.GetRir() != rva.RIR {
+				err = fmt.Errorf(
+					"Expected perspective %q (%q) but got reply from %q (%q) - misconfiguration likely", rva.Perspective, rva.RIR, res.GetPerspective(), res.GetRir(),
+				)
+				responses <- &response{rva.Address, rva.Perspective, rva.RIR, res, err}
+				return
+			}
+			responses <- &response{rva.Address, rva.Perspective, rva.RIR, res, err}
+		}(va.remoteVAs[i])
+	}
+
+	required := remoteVACount - va.maxRemoteFailures
+	var passed []string
+	var failed []string
+	var passedRIRs = map[string]struct{}{}
+	var firstProb *probs.ProblemDetails
+
+	for resp := range responses {
+		var currProb *probs.ProblemDetails
+
+		if resp.err != nil {
+			// Failed to communicate with the remote VA.
+			failed = append(failed, resp.perspective)
+
+			if core.IsCanceled(resp.err) {
+				currProb = probs.ServerInternal("Secondary validation RPC canceled")
+			} else {
+				va.log.Errf("Operation on remote VA (%s) failed: %s", resp.addr, resp.err)
+				currProb = probs.ServerInternal("Secondary validation RPC failed")
+			}
+		} else if resp.result.GetProblem() != nil {
+			// The remote VA returned a problem.
+			failed = append(failed, resp.perspective)
+
+			var err error
+			currProb, err = bgrpc.PBToProblemDetails(resp.result.GetProblem())
+			if err != nil {
+				va.log.Errf("Operation on Remote VA (%s) returned malformed problem: %s", resp.addr, err)
+				currProb = probs.ServerInternal("Secondary validation RPC returned malformed result")
+			}
+		} else {
+			// The remote VA returned a successful result.
+			passed = append(passed, resp.perspective)
+			passedRIRs[resp.rir] = struct{}{}
+		}
+
+		if firstProb == nil && currProb != nil {
+			// A problem was encountered for the first time.
+			firstProb = currProb
+		}
+
+		// To respond faster, if we get enough successes or too many failures, we cancel remaining RPCs.
+		// Finish the loop to collect remaining responses into `failed` so we can rely on having a response
+		// for every request we made.
+		if len(passed) >= required && len(passedRIRs) >= requiredRIRs {
+			cancel()
+		}
+		if len(failed) > va.maxRemoteFailures {
+			cancel()
+		}
+
+		// Once all the VAs have returned a result, break the loop.
+		if len(passed)+len(failed) >= remoteVACount {
+			break
+		}
+	}
+
+	if len(passed) >= required {
+		if len(passedRIRs) >= requiredRIRs {
+			return summarizeMPIC(passed, failed, passedRIRs), nil
+		}
+		return summarizeMPIC(passed, failed, passedRIRs), probs.Unauthorized(
+			"During secondary validation: validation could not be corroborated by enough distinct RIRs",
```

Comment:
From a user perspective, I think this error message would be very confusing. We should treat this case the same as if we failed the reach the threshold for total successes, and return the first problem. I'm thinking something like:

```
if len(passed) >= required && len(passedRIRs) >= requiredRIRs {
    return summarizeMPIC(passed, failed, passedRIRs), nil
}
// This should never happen. If we didn't meet the thresholds above
// we should have seen at least one error.
if firstProb == nil {
  firstProb = probs.ServerInternal("Too few remote RPC results")
} else {
  firstProb.Detail = fmt.Sprintf("During secondary validation: %s", firstProb.Detail)
}
return summarizeMPIC(passed, failed, passedRIRs), probs.ServerInternal("Too few remote RPC results")
```

Note that this restructures the existing `if` / `else if` / `else` code a bit, but I think that winds up being necessary because we now have two criteria for success and it would be clumsy to put their inverses in the `else if` clause.

Reply:
Aaron left a similar comment concerning this. I believe I've arrived at a solution that should satisfy both of you:

```go
	if len(passed) >= required && len(passedRIRs) >= requiredRIRs {
		return summarizeMPIC(passed, failed, passedRIRs), nil
	} else if len(passed) >= required && firstProb == nil {
		// This should never happen. If we didn't meet the thresholds above we
		// should have seen at least one error.
		return summarizeMPIC(passed, failed, passedRIRs), probs.Unauthorized(
			"During secondary validation: validation could not be corroborated by enough distinct RIRs",
		)
	} else if len(failed) > va.maxRemoteFailures {
		firstProb.Detail = fmt.Sprintf("During secondary validation: %s", firstProb.Detail)
		return summarizeMPIC(passed, failed, passedRIRs), firstProb
	} else {
		// This condition should not occur - it indicates the passed/failed counts
		// neither met the required threshold nor the maxRemoteFailures threshold.
		return summarizeMPIC(passed, failed, passedRIRs), probs.ServerInternal("Too few remote RPC results")
	}
```